### PR TITLE
Provide option to retain punctuation tokens

### PR DESCRIPTION
**Pull request by jayaddison**
In the same use case as #15, I'd like to receive punctuation characters in the token output stream.  Although these tokens will not be stored as terms in an index, it can be useful during [highlighting](https://github.com/openculinary/hashedixsearch/blob/4acec272d06d8d0923c2ca3e73522f7c5cf306d7/hashedixsearch/search.py#L141-L186) to reconstruct the original input text verbatim while also using consistent non-punctuation term stemming.

~~To be cautious I've added test coverage not just over the new functionality, but also over some existing ngram-related functionality to make sure it behaves consistently.~~

(apologies for not realizing that I should mention this at the same time during the previous pull request :))

Edit: the ngram-related tests have now been removed, given more confidence that they're not required.

### CODE REVIEW

**Comment by jayaddison** [comment]
(FWIW - this should be the only additional change I need at the moment; I promise there won't be follow-up pull request churn today!)

File: hashedindex/textparser.py [hunk 1]
```diff
@@ -92,13 +96,14 @@ def word_tokenize(text, stopwords=_stopwords, ngrams=None, min_length=0, ignore_
...
-    text = re.sub(_re_punctuation, '', text)
+    text = text if retain_punctuation else re.sub(_re_punctuation, '', text)
     text = text if retain_casing else text.lower()
 
     matched_tokens = re.findall(_re_token, text)
     for tokens in get_ngrams(matched_tokens, ngrams):
         for i in range(len(tokens)):
-            tokens[i] = tokens[i].strip(punctuation)
+            if not retain_punctuation:
+                tokens[i] = tokens[i].strip(punctuation)
```
Comment (jayaddison) [inline hashedindex/textparser.py:106]: @MichaelAquilina Could it make sense to re-use `_re_punctuation` for this step?  I'm wondering if there may be any cases where `string.punctuation` and `_punctuation` may end up differing slightly, leading to unusual behaviour.

Comment (jayaddison) [reply hashedindex/textparser.py:106]: I suppose this is for *within-token* punctuation characters (i.e. `pre-heat` -> `preheat`) - that's certainly a slightly different situation to splitting token boundaries.

**Comment by jayaddison** [comment]
:hourglass_flowing_sand: There's definitely some subtletly in punctuation handling, so I'm going to spend a little time investigating and will likely introduce some additional test cases.

**Comment by jayaddison** [comment]
Getting closer.  One more subtle issue around the `_punctuation_exceptions` to track down.

**Comment by jayaddison** [comment]
Ah, of course.  `strip` only replaces leading and trailing characters from a string.  This explains some things :)

**Comment by jayaddison** [comment]
Ok, having done a bit more investigation and testing of the within-token-punctuation scenarios, this should now be ready for review.

**Comment by jayaddison** [comment]
Q: How would a caller of `word_tokenize` know how to re-assemble tokens, and how much whitespace to combine between them?

**Comment by MichaelAquilina** [comment]
> (FWIW - this should be the only additional change I need at the moment; I promise there won't be follow-up pull request churn today!)

Dw about that :)

> How would a caller of word_tokenize know how to re-assemble tokens, and how much whitespace to combine between them?

Right now, there is no perfect way to reconstruct it. You would have to work under the assumption that everything was split as the same token and rejoin (e.g. `' '.join(tokens)`). Might be an improvement to also output the list of whitespace tokens parsed if you find it useful somehow.

File: tests/test_parser.py [hunk 2]
```diff
@@ -131,6 +136,18 @@ def test_retains_casing(self):
...
+        assert list(textparser.word_tokenize(
+            text='who, where? (question!)',
+            retain_punctuation=True
+        )) == [('who', ), (',', ), ('where',), ('?', ), ('(',), ('question',), ('!',), (')',)]
+
+    def test_retains_punctuation_within_tokens(self):
+        assert list(textparser.word_tokenize(
+            text='is the oven pre-heated?',
+            retain_punctuation=True
+        )) == [('is', ), ('the', ), ('oven',), ('pre-heated', ), ('?',)]
```
Comment (MichaelAquilina) [inline tests/test_parser.py:149]: why do we dinstinguish between `-` and other punctuation here?

File: hashedindex/textparser.py [hunk 3]
```diff
@@ -83,6 +87,9 @@ def word_tokenize(text, stopwords=_stopwords, ngrams=None, min_length=0, ignore_
 
     Generated tokens are lowercased by default; the retain_casing flag can
     be set to True to retain upper/lower casing from the original text.
+
+    When retain_punctuation is set to False (default), punctuation characters
+    are removed from the input text.
```
Comment (MichaelAquilina) [inline hashedindex/textparser.py:92]: This is slightly hard to read - maybe take a similar format for what you wrote for `retain_casing`?

File: tests/test_parser.py (hunk 2 above)
Comment (jayaddison) [reply tests/test_parser.py:149]: The difference (and it's debatable whether this is the correct way to handle things) is that the `-` here is punctuation encountered inside a word token.  Currently that value (and also `\` and `/`) are handled as special cases (see https://github.com/MichaelAquilina/hashedindex/pull/16/files#diff-bca7df3b022757454ef625ad3579a5b5R27-L29)

**Comment by jayaddison** [comment]
> > (FWIW - this should be the only additional change I need at the moment; I promise there won't be follow-up pull request churn today!)
> 
> Dw about that :)
> 
> > How would a caller of word_tokenize know how to re-assemble tokens, and how much whitespace to combine between them?
> 
> Might be an improvement to also output the list of whitespace tokens parsed if you find it useful somehow.

This is my preferred solution, I think.  I don't want to proliferate the number of `word_tokenize` arguments, but I wonder if this should also be a configurable option (`tokenize_whitespace` or similar).

File: hashedindex/textparser.py [hunk 4]
```diff
@@ -92,15 +99,13 @@ def word_tokenize(text, stopwords=_stopwords, ngrams=None, min_length=0, ignore_
...
 
     text = re.sub(re.compile('\'s'), '', text)  # Simple heuristic
-    text = re.sub(_re_punctuation, '', text)
+    text = text if retain_punctuation else re.sub(_re_punctuation, '', text)
     text = text if retain_casing else text.lower()
 
     matched_tokens = re.findall(_re_token, text)
     for tokens in get_ngrams(matched_tokens, ngrams):
         for i in range(len(tokens)):
-            tokens[i] = tokens[i].strip(punctuation)
```
Comment (jayaddison) [inline hashedindex/textparser.py:101]: :hourglass_flowing_sand: I think this line should be restored.  I'll add some before/after test coverage to illustrate why.

Comment (jayaddison) [reply hashedindex/textparser.py:101]: :heavy_check_mark: Good news - no need for additional coverage here; this case was already [covered by a test](https://github.com/MichaelAquilina/hashedindex/blob/783c7adfaf11e73fe40bc3535b36f029cc28b8b2/tests/test_parser.py#L105-L108) prior to the commit this changeset branches from.  The test name could possibly be updated; PR #19 is available to optionally make that change if it seems worthwhile.

**Comment by jayaddison** [comment]
@MichaelAquilina Oops.  This PR should remain open -- my hyperlink from #19 to a comment in here caused it to get closed automatically.  That's a good GitHub edge case to know about.

**Comment by jayaddison** [comment]
Ok if we re-open this one @MichaelAquilina?

**Comment by MichaelAquilina** [comment]
Re-opened the PR. Will try find some time to review everything at some point!

**Comment by jayaddison** [comment]
No rush, thank you :)

**Comment by jayaddison** [comment]
@MichaelAquilina I'm a bit torn on this - while I don't want to apply pressure, and you should definitely take time to review this and #17 fully, it would be an excellent close to the week over here if these changes could slot in.

Please let me know if I can do anything to help explain the changes further and/or ease the review process.

File: tests/test_parser.py (hunk 2 above)
Comment (MichaelAquilina) [reply tests/test_parser.py:149]: This should probably be documented somewhere (I cant remember my justification for having these exceptions as its been a long time!)

**Comment by MichaelAquilina** [comment]
> @MichaelAquilina I'm a bit torn on this - while I don't want to apply pressure, and you should definitely take time to review this and #17 fully, it would be an excellent close to the week over here if these changes could slot in.
> Please let me know if I can do anything to help explain the changes further and/or ease the review process.

Sorry about that - its mostly just a case of me finding the time and energy in the evening after work!

**Pull request merged**